### ISSUE TITLE
the outcome seems not good

### ISSUE THREAD

**Issue by YeHaijia** [HRNet/Higher-HRNet-Human-Pose-Estimation#17]

**Comment by Daniil-Osokin** [comment]
Hi,
I believe this is due to drawing all found keypoints, even with low score. Try to draw keypoints, in which network more or less confident, e.g. which have `>= 0.1` value in heatmap.

**Comment by plmsmile** [comment]
> Hi,
> I believe this is due to drawing all found keypoints, even with low score. Try to draw keypoints, in which network more or less confident, e.g. which have `>= 0.1` value in heatmap.
> 
> > P.S. This mode is useful for results submission, because COCO dataset expects to have 17 keypoints for any person, even if it truncated as in the 3rd image. And some keypoints with low confidence may be not so far from the target, so still positively contribute to accuracy metric, thus improving final score.

I met the same problem and the result looks very bad. 

- Where can I set the threshold? (I tried DETECTION_THRESHOLD=0.5, may not this param)
- The results showed that may threshold is not the only bad reason.

![image](https://user-images.githubusercontent.com/17981649/72051521-6c671b80-32fe-11ea-94cd-0fd99564bac9.png)
![image](https://user-images.githubusercontent.com/17981649/72051627-9ddfe700-32fe-11ea-8ca3-44bf493364c6.png)
![image](https://user-images.githubusercontent.com/17981649/72051750-cc5dc200-32fe-11ea-8393-abb17a4c59f0.png)

**Comment by Daniil-Osokin** [comment]
I cannot find the threshold for keypoints in config file. Actually, keypoint confidence is the 3<sup>rd</sup> value in [`final_results`](https://github.com/HRNet/Higher-HRNet-Human-Pose-Estimation/blob/master/tools/valid.py#L188) (in 3<sup>rd</sup> dim). Suppose you use `save_valid_image` for visualization, so here is the diff:

```
diff --git a/lib/utils/vis.py b/lib/utils/vis.py
index f351a0a..f4b169e 100755
--- a/lib/utils/vis.py
+++ b/lib/utils/vis.py
@@ -20,12 +20,13 @@ from dataset import VIS_CONFIG
 def add_joints(image, joints, color, dataset='COCO'):
     part_idx = VIS_CONFIG[dataset]['part_idx']
     part_orders = VIS_CONFIG[dataset]['part_orders']
+    kpt_threshold = 0.1
 
     def link(a, b, color):
         if part_idx[a] < joints.shape[0] and part_idx[b] < joints.shape[0]:
             jointa = joints[part_idx[a]]
             jointb = joints[part_idx[b]]
-            if jointa[2] > 0 and jointb[2] > 0:
+            if jointa[2] > kpt_threshold and jointb[2] > kpt_threshold:
                 cv2.line(
                     image,
                     (int(jointa[0]), int(jointa[1])),
@@ -36,7 +37,7 @@ def add_joints(image, joints, color, dataset='COCO'):
 
     # add joints
     for joint in joints:
-        if joint[2] > 0:
+        if joint[2] > kpt_threshold:
             cv2.circle(image, (int(joint[0]), int(joint[1])), 1, color, 2)
 
     # add link
```

And result:
![result_valid_11](https://user-images.githubusercontent.com/2491976/72096017-73a41e80-332a-11ea-8ff2-8b4e9bb337a9.jpg)

**Comment by leoxiaobin** [comment]
As @Daniil-Osokin said, you should add a threshold to filter the low score points.

**Comment by hothothot2000** [comment]
The strange links mentioned in my post represents very long lines linked to joints out of image.
I have tried kpt_threshold, this solved some of the strange links.
But, a few strange links still appear even I set kpt_threshold=0.4 .
And what's more, single person sometimes was drawed links in several colors.

I think that means this person has been counted repeatedly as different persons.
As a proof of my guess, sometimes there are 80+ persons detected in final_results of frame of my video with only 3 ~ 7 persons inside.

Maybe we need set a person_threshold? And Where to set this will make sense?

**Comment by Daniil-Osokin** [comment]
You can play with parameters for your specific use case. Addition of `person_threshold` is an option, e.g. you can sum all keypoints confidence (`joint[2]` in the code above) and draw a person if this sum is higher than some threshold. Or you can count number of keypoints with high confidence and draw a pose if it has significant number of such keypoints (e.g.more than a half of total keypoints number).

**Comment by hothothot2000** [comment]
> You can play with parameters for your specific use case. Addition of `person_threshold` is an option, e.g. you can sum all keypoints confidence (`joint[2]` in the code above) and draw a person if this sum is higher than some threshold. Or you can count number of keypoints with high confidence and draw a pose if it has significant number of such keypoints (e.g.more than a half of total keypoints number).


Thanks for your advice, I think your method will probably work.
But, this is only one section of the total problem.
The other section is -- the detected 80+ persons and their joints in the frame of actual 7 persons, will consume much more caculate time.
The "Inf_time" of these frames is really about 10 times or more than "Inf_time" of those frames not over caculated.
May there be any way to save these time on over caculation?

**Comment by Daniil-Osokin** [comment]
The tricks above can deal with this. You just need to apply them at the time of poses construction, not at the time of result visualization. Just set higher threshold to keypoints values in heatmaps before grouping into the poses, and the time will be spend only on poses, in which network is strongly confident.

**Comment by hothothot2000** [comment]
![image](https://user-images.githubusercontent.com/25924605/80795259-90f94300-8bce-11ea-972e-98289d3afbf1.png)

> The tricks above can deal with this. You just need to apply them at the time of poses construction, not at the time of result visualization. Just set higher threshold to keypoints values in heatmaps before grouping into the poses, and the time will be spend only on poses, in which network is strongly confident.

Thanks a lot.
This trick did help to remove majority of the unreasonable locations of joints when we set kpt_threshold=0.3 before grouping.
But, minority still stayed with confidance >0.3, a few even stayed with confidance >0.4.
And a collateral negative effect is, some resonable joints also had been discarded if kpt_threshold was set to 0.4 or higher.

Above from that, we occasionally found some joints-pair links are incomprehensible.
For example, A's knee was linked to B's ankle when they are not adjacent, or even apart by a considerably far distance.
thus we are really confused by these mismatched links.

May there be more hint for these?

**Comment by Daniil-Osokin** [comment]
> A's knee was linked to B's ankle

This behavior seems buggy for me (but possibly this is a feature to maximize leaderboard score). In general, parts of different persons should not be linked together. To get rid of far links you may try to estimate mean size of links between keypoints of the same person and filter out outliers (ones, which longer than twice mean size). @hothothot2000 BTW, this is just a repository of research work with own pros and cons. To apply its results in production solution, consider using consulting specialists. It can really save the time and give necessary quality level.

**Comment by YinXiaojieCSDN** [comment]
> I use pose_higher_hrnet_w32_512.pth to test some my own images.But the outcome seem not good?what's wrong?
> ![image](https://user-images.githubusercontent.com/18445398/71968665-5430c700-3240-11ea-85b8-498c4727e7a7.png)
> ![image](https://user-images.githubusercontent.com/18445398/71968700-627ee300-3240-11ea-8009-4f3fa944187b.png)
> ![image](https://user-images.githubusercontent.com/18445398/71968762-80e4de80-3240-11ea-8016-019b4eda45f1.png)

can you give me yours enviroment list(or pip list)?,I be stracked when i valid.

**Comment by Daniil-Osokin** [comment]
Will [requirements.txt](https://github.com/HRNet/HigherHRNet-Human-Pose-Estimation/blob/master/requirements.txt#L1) work for you?

**Comment by joaqo** [comment]
Hi @Daniil-Osokin, do you mind posting the diff of the code you are using to run inference on a custom image and draw/filter the keypoints?

**Comment by Daniil-Osokin** [comment]
Hi, I've run [valid.py](https://github.com/HRNet/HigherHRNet-Human-Pose-Estimation/blob/master/tools/valid.py) on COCO validation data.

**Comment by joaqo** [comment]
Thanks, tried it on a simple 6-person image on got perfect results, but then I tried it on a video, in a much more complicated scene, and these are the results I'm getting: 
<img width="1916" alt="image" src="https://user-images.githubusercontent.com/14966348/84089125-982d2100-a9c4-11ea-8740-be4f611220dc.png">

I used the following command to run it: 
```python tools/valid.py     --cfg experiments/coco/higher_hrnet/w32_512_adam_lr1e-3.yaml     TEST.MODEL_FILE models/pytorch/pose_coco/pose_higher_hrnet_w32_512.pth     TEST.FLIP_TEST False```

So I can validate that there is some issue with the pose decoding on hard examples.

**Comment by Daniil-Osokin** [comment]
Actually there are some nice poses in the output. To get rid of mispredicted results a post-processing (which discussed above) is required.

**Comment by joaqo** [comment]
Gotcha, I tried filtering according to the diff you posted with regards to vis.py, but using a 0.3 threshold, and the results were less noisy, but not ideal.

<img width="1915" alt="image" src="https://user-images.githubusercontent.com/14966348/84186237-a59ae780-aa66-11ea-9895-2a3137802596.png">

I understand that filtering at the heatmap tensor level will probably filter the erroneous association between different people, still most of the people in the image aren't getting detected at all. Am I using a model version which is too small? Maybe a model trained on CrowdPose would better fit this particular sample?

Thanks for the help, and for open sourcing the model!

**Comment by Daniil-Osokin** [comment]
I would say the first thing for this image to get better results is to increase network input resolution (possibly set scale 2 (or even 4) [here](https://github.com/HRNet/HigherHRNet-Human-Pose-Estimation/blob/master/experiments/coco/higher_hrnet/w48_640_adam_lr1e-3.yaml#L109)).

Thanks, however this work is not mine, I'm just answering on some questions. I've got the result below (just ran on image above) with [other](https://github.com/Daniil-Osokin/lightweight-human-pose-estimation.pytorch) method, so this should also work.
![lightweight_openpose](https://user-images.githubusercontent.com/2491976/84200618-ca677d00-aaaf-11ea-8a26-7e24dcd88365.jpg)

**Comment by joaqo** [comment]
Hi @Daniil-Osokin, one question, the repo you posted reported 40mAP, which I think would be much worse accuracy wise than the image you posted which seems to be really good. Is it possible you actually used one of the newer neighbour projects from that group? I would try them but they seem to require OpenVINO. Thanks!

**Comment by Daniil-Osokin** [comment]
You can reproduce this result using `--height-size=1024`.

**Comment by seekFire** [comment]
@Daniil-Osokin @plmsmile @hothothot2000 
I have found there has a fatal mistake in this frame code, details in [my issue](https://github.com/HRNet/HigherHRNet-Human-Pose-Estimation/issues/83)

**Comment by seekFire** [comment]
@Daniil-Osokin @plmsmile @hothothot2000 @YeHaijia 
I have found there has a fatal mistake in this frame code, details in [my issue](https://github.com/HRNet/HigherHRNet-Human-Pose-Estimation/issues/83), I think the mistake may explain the case shown above intuitively.

**Comment by seekFire** [comment]
@Daniil-Osokin @plmsmile @hothothot2000 @YeHaijia @leoxiaobin 
After checking the code, I think I find out the reason for the bad outcome:
One of the transforms used in training stage: affine Transform, take the responsibility of mapping the original coordinates into the heatmap coordinates (see the code of function **build_transforms** in /lib/dataset/transforms/build.py), it only used in training dataset, so the training stage runs OK (see the code of function **make_dataloader** in /lib/dataset/build.py), but it not used in testing stage(see the code of function **make_test_dataloader** in /lib/dataset/build.py, the **transforms** is set to be **None**), 
so the test result may have such mistake.

**Comment by DearPerpetual** [comment]
> You can reproduce this result using `--height-size=1024`.

What is the reason why I used this network to train my dataset with only 5 key points, and my training phase was normal, but the results during testing were poor and I couldn't look directly at them?
